src/main/java/com/google/devtools/build/lib/vfs/AbstractFileSystem.java

```diff
@@ -48,35 +52,65 @@ protected InputStream getInputStream(Path path) throws IOException {
     // This loop is a workaround for an apparent bug in FileInputStream.open, which delegates
     // ultimately to JVM_Open in the Hotspot JVM.  This call is not EINTR-safe, so we must do the
     // retry here.
+    InputStream stream = null;
     for (; ; ) {
       try {
-        return createFileInputStream(path);
+        stream = createFileInputStream(path);
       } catch (FileNotFoundException e) {
         if (e.getMessage().endsWith("(Interrupted system call)")) {
           continue;
         } else {
           throw e;
         }
+      } catch (AccessDeniedException e) {
+        throw new IOException(e.getMessage() + " (Permission denied)", e);
       }
+      break;
     }
+
+    // On some platforms, whether a path can be read as a file won't be checked until
+    // the first read().
+    //
+    // http://mail.openjdk.java.net/pipermail/nio-dev/2014-December/002877.html
+    //
+    // The underlying stream might not be seekable, and wrapping in a BufferedInputStream
+    // causes test failures in the integration suite, so create our own little one-byte
+    // buffer here.
+    byte[] buf = new byte[1];
+    int buflen = 0;
+    try {
+      buflen = stream.read(buf, 0, 1);
+    } catch (IOException e) {
+      if (e.getMessage().equals("Is a directory")) {
+        throw new IOException(path.toString() + " (Is a directory)", e);
```

Comment:
Why not just `throw e;`? (granted, this one has a slightly nicer stack trace)

Reply:
The error message doesn't have the filename in it, which makes it (1) unhelpful for users and (2) fail a test covering this exact case.